algorithms/src/snark/marlin/ahp/prover/round_functions/second.rs

```diff
@@ -15,360 +15,178 @@
+
+                    if rowcheck.coeffs.iter().any(|&c| c != F::zero()) {
+                        instance_lhs += &(&rowcheck * instance_combiner);
+                    }
+
+                    // Let R = largest_domain;
+                    // Let R_i = domain;
+                    // Let v_R := R.vanishing_polynomial();
+                    // Let v_R_i := R_i.vanishing_polynomial();
+                    // Let s_i := R.selector_polynomial(R_i) = (v_R / v_R_i) * (R_i.size() / R.size());
+                    // Let c_i := circuit combiner
+                    // Let lhs_i := circuit_specific partial left hand size of the sumcheck, see protocol docs for more details
+
+                    // Instead of just multiplying each lhs_i by `s_i*c_i`, we reorder the sumcheck to cancel out division by v_R
+                    // This removes a mul and div by v_R operation over each circuit's (max_constraint_domain - constraint_domain)
+                    // Substituting in s_i, we get that lhs_i * s_i / v_R = lhs_i / v_K * (R_i.size() / R.size());
+
+                    let selector_rowcheck_time =
+                        start_timer!(|| format!("Compute selector rowcheck for {}", _circuit_id));
+
+                    let (mut h_0_i, remainder) = instance_lhs
+                        .divide_by_vanishing_poly(constraint_domain)
+                        .ok_or(anyhow::anyhow!("could not divide by v_R"))?;
+                    assert!(remainder.is_zero());
+                    let multiplier =
+                        circuit_combiner * constraint_domain.size_as_field_element * max_constraint_domain_inv;
+                    cfg_iter_mut!(h_0_i.coeffs).for_each(|c| *c *= multiplier);
+
+                    end_timer!(selector_rowcheck_time);
+
+                    Ok::<_, anyhow::Error>(h_0_i)
+                });
+            }
         }
 
-        let mask_poly = state.first_round_oracles.as_ref().unwrap().mask_poly.as_ref();
-        assert_eq!(MM::ZK, mask_poly.is_some());
-        let mask_poly = &mask_poly.map_or(DensePolynomial::zero(), |p| p.polynomial().into_dense());
-        let (mut h_1_mask, mut xg_1_mask) = mask_poly.divide_by_vanishing_poly(max_constraint_domain).unwrap();
-        h_1_sum += &core::mem::take(&mut h_1_mask);
-        xg_1_sum += &core::mem::take(&mut xg_1_mask);
+        let h_sum_time = start_timer!(|| "AHP::Prover::SecondRound h_sum");
+        let h_sum: DensePolynomial<F> =
+            cfg_reduce!(cfg_into_iter!(job_pool.execute_all()), || Ok(DensePolynomial::zero()), |a, b| {
+                if let Ok(mut b) = b {
+                    if let Ok(mut a) = a {
+                        if b != DensePolynomial::zero() {
+                            a += &std::mem::take(&mut b);
+                        }
+                        return Ok(a);
+                    }
+                }
+                Err(anyhow::anyhow!("Could not sum h instances"))
```

Comment:
```suggestion
                a.and_then(|mut a| b.as_mut().map(|b| a + &std::mem::take(b)))
              
```

Reply:
That. Is. beautiful. 1da3db9e2
FYI DensePolynomial doesn't have a `+` API and I assumed this was intentioned design